README.md

```diff
@@ -1,10 +1,39 @@
+# jit-cleanup
 
+This is python web application which is to clean up [jit-access](https://github.com/GoogleCloudPlatform/jit-access) expired permission.
+
+
+## Background
+There is an existing solution published on the GCP architecture site which provides just in time access, open sourced as [jit-access](https://github.com/GoogleCloudPlatform/jit-access). It leverages an IAP protected application and templated IAM permissions to grant users time based privileged access. It provides a robust solution that includes logging, self approval, multi-party approval and variable time based access.
+Under the hood it creates specific user permissions with IAM conditions to enforce time based access. The tool is additive, and each new request generates a user IAM binding with a new condition, unless the variable PURGE_EXISTING_TEMPORARY_BINDINGS  (currently undocumented)  is set in the app.yaml. In that scenario only a single conditional access binding per user is maintained.
+
+Here is [how to setup](https://cloud.google.com/architecture/manage-just-in-time-privileged-access-to-project) JIT in GCP.
+
+In general purging temporary bindings option will avoid hitting limits on max conditional bindings, but there are some challenges to implement this sometimes, including:
+
+- IaC drift: will occur when temporary permissions are implemented using direct API calls instead of IaC. This can lead to changes in the project IAM binding configuration, which can be an operational risk.
+- Compliance issue: Any JIT conditional bindings may trigger a failure in PaC unless we can exclude them.
+
+## Solution
+The easiest way to handle expired user based bindings is to remove the conditional access after expiry. This can be accomplished by running a periodic cleanup task to remove expired bindings.
+Instead of storing state in a database, we can use PubSub to store events until they are processed. This allows multiple components to act on published events, not just for removing conditional bindings. This provides more flexibility and scalability.
+
+
+
+## Deployment setup
+### Before You Start
+Make sure you have the [JIT developed in your GCP environment](https://cloud.google.com/architecture/manage-just-in-time-privileged-access-to-project)
+
+
+### Deploy code
 ```shell
 REGION=asia-southeast1
-PROJECT_ID=ethanhanjoonix-proj1
+PROJECT_ID=xxx
 SERVICE=jit-cleanup
-SERVICE_ACCOUNT_EMAIL=jitaccess@ethanhanjoonix-proj1.iam.gserviceaccount.com
+SERVICE_ACCOUNT_EMAIL=jitaccess@xxx.iam.gserviceaccount.com
 REPO_NAME=jit-repo
+PUBSUB_TOPIC_NAME=projects/xxx/topics/jit-access
+PUBSUB_SUBSCRIPTION_PATH=projects/xxx/subscriptions/jit-sub
```

Comment:
hmm, actually you need to create a subscription with a filter to ensure you receive just the messages you're interested in. Seems like you're just using the default. Please look at https://cloud.google.com/pubsub/docs/subscription-message-filter#creating_subscriptions_with_filters

Reply:
creating filter is specified in gcloud command. it should be decoupled from code?

Reply:
Yes it will need to be decoupled from the code. As this is a scheduled based task if you created/deleted when the function ran then you'd only every pick up messages sent whilst it was running. A subscription needs to be present to collect the items for processing.

Reply:
sry, I may not understand well, do I need to specify the create subscription with filter gcloud command in the readme doc? cus currently I only provided create subscription gcloud command without filter in readme doc.